### ISSUE TITLE
meta-chart.com

### ISSUE THREAD

**Issue by gotitbro** [uBlockOrigin/uAssets#618]
Anti adblock warning is displayed on https://www.meta-chart.com. They are using Admiral's (`https://getadmiral.com`) anti adblock technology.

**Error URL:** https://www.meta-chart.com

Arrived at this domain from https://github.com/easylist/easylist/commit/a4d380ad1a3b33a0fab679a1a8c5a791321622b3#commitcomment-23587399

@ryanbr There removed an Admiral ad domain from EasyList citing a DMCA request which has not been posted publicly till now. The filter that has been removed is:
`||functionalclam.com^$third-party`

This ad domain is being used on this website (meta-chart.com) among others to display the anti adblock warning.

_**Note:** Using uBO 1.13.8 with the default filters and settings._

**Comment by okiehsch** [comment]
Copyright law does not protect domain names, so how can this result in a valid DMCA request?
Anyway
````
meta-chart.com##script:inject(setTimeout-defuser.js, "ubo", 300)
meta-chart.com##script:inject(setTimeout-defuser.js, [native code], 15000)
````
fixes your issue.

**Comment by uBlock-user** [comment]
DMCA has nothing to do with filters and filterlists and can't be used to compel someone to remove a filter from filterlist. This is laughable at best.

**Comment by gorhill** [comment]
Solution does not seem to work for Nightly + uBO 1.13.9b8. I wonder whether this is a Nightly bug causing scriptlets to not be injected, but I can defuse properly with `bab-defuser.js` on `blockadblock.com`.

**Comment by gorhill** [comment]
I got it to work with `meta-chart.com##script:inject(setTimeout-defuser.js, setTimeout)`.

Anybody else can reproduce the issue with Nightly?

**Comment by gorhill** [comment]
I created a new generic scriptlet to make it easier to abort according to current inline script rather than arguments passed to specific function calls.

For example, in the current case, this works without the need for `setTimeout-defuser.js`:

    meta-chart.com##script:inject(abort-current-inline-script.js, Promise)

`Promise` is the entry point, inside which the scriptlet will look at the current inline script to find out if it matches the 3rd argument. No 3rd argument means always matches. In plain English words, the filter above means: "If `window.Promise` is accessed directly from within any inline script, then abort execution that script". A 3rd argument would be used to further narrow by requiring a match with the content of the current inline script.

**Comment by moshpirit** [comment]
If I understood this right, what DMCA does is forcing websites like Easylist to delete some others so they can have ads. So, when we use adblocks like ublock origin, it doesn't download "website X" from Easylist, therefore, it allows this ad. Am I right? then, why not allowing the user to export our current list of websites and just add more with the time so "website X" is in our filters and can be blocked?

**Comment by okiehsch** [comment]
The DMCA-complaint states that:
 The code in question attempts to circumvent copyright access controls to copyrighted content on the site listed in the title of this [commit](https://github.com/easylist/easylist/commit/1ba8d4afeec6d562a5871fc7504c756e4b2bd5bc).

So in essence Admiral is arguing that `||functionalclam.com^$third-party` violates the right of the website
`mathwarehouse.com`, to only allow access to their copyrighted content if you also connect to `functionalclam.com`.

I do not know what you mean by 

> why not allowing the user to export our current list of websites and just add more with the time so "website X" is in our filters and can be blocked?

you can already import any list you like to your uBO setup and add any website to your own filter list.

**Comment by gotitbro** [comment]
@SMed79 @okiehsch @gorhill I think this might be the right place to ask. So, what came out of the DMCA controversy? I haven't been able to follow it after the initial day.

**Comment by uBlock-user** [comment]
@gotitbro Easylist caved in and removed the filter which was what Admiral wanted after all, however this action caused [Streisand_effect](https://en.wikipedia.org/wiki/Streisand_effect) on a widespread on Github and in other places and functionalclam was blocked on multiple repositories, nothing BarbBlock, an extension built to fight DMCA was created.

**Comment by mapx-** [comment]
it seems `meta-chart.com##script:inject(abort-current-inline-script.js, Promise)` does not work anymore

**Comment by gorhill** [comment]
The filter still work for me on that site. If I un-check "uBlock filters", I get the anti-blocker warning. If I enable "uBlock filters", no warning.

**Comment by mapx-** [comment]
I tested chrome 61, 62 + uBo 1.14.8 + default filters (included updated ublock filters), firefox 57 (uBo 1.14.11rc8). I get always the warning msg.

**Comment by okiehsch** [comment]
With Chrome 61, Firefox 56 and uBO 1.14.8, or  1.14.11rc8, all works fine on my end.

**Comment by mapx-** [comment]
I used 4 different browsers, same setup in uBo, still getting the warning. No idea what's happening.

**Comment by mapx-** [comment]
Tested also in opera, uBo 1.13.8, same behaviour.

**Comment by okiehsch** [comment]
I tried with two different machines with different OSes, Windows 1ß and Linux Manjaro, works woth both
setups on my end.
Does
````
meta-chart.com##script:inject(setTimeout-defuser.js, "ubo", 300)
meta-chart.com##script:inject(setTimeout-defuser.js, [native code], 15000)
````
work for you?

**Comment by mapx-** [comment]
So: your filters are working in chrome 61, 62, opera.
Don't work in FF 57 + uBo 1.4.11rc8 (probably because uBo does not read anymore resources.txt ? )

**Comment by okiehsch** [comment]
No, Firefox needs another `setTimeout-call` defused, seems new to me, anyway
````
meta-chart.com##script:inject(setTimeout-defuser.js, "ubo", 300)
meta-chart.com##script:inject(setTimeout-defuser.js, [native code], 15000)
meta-chart.com##script:inject(setTimeout-defuser.js, [native code], 0)
````
works for me and I will have to add 
`example.com##script:inject(setTimeout-defuser.js, [native code], 0)` to uBO-filters, because they do not work, using Firefox, right now on my end, example link:
````
http://www.post-gazette.com/business/money/2017/09/29/More-parents-are-finding-ways-to-save-money-for-college/stories/201709250142
````

**Comment by mapx-** [comment]
yes, the above 3 filters are working for me too in FF.

**Comment by okiehsch** [comment]
Ther is still the question why 
`meta-chart.com##script:inject(abort-current-inline-script.js, Promise)` does not work for you.
Does `meta-chart.com##script:inject(abort-current-inline-script.js, setTimeout)` work for you, 
if not can you look at the source-code of `meta-chart.com` to  check if we both get served the
same document.
Search for `window.promise`, you should get two hits.

**Comment by mapx-** [comment]
yes, twice, see the inline script here:
https://mega.nz/#!WcwkADha!EBXXj3CLeMvAE4SD1MhkzrAbFGqf6ZFqt1S7z3qBM3c

**Comment by mapx-** [comment]
and this filter
`meta-chart.com##script:inject(abort-current-inline-script.js, setTimeout)`

does not work for me (in chrome)

**Comment by okiehsch** [comment]
Well, then I don't know why it does not work for you, they both work on my end, sorry.
I suppose `meta-calculator.com` also does not work on your end, uBO also uses
`##script:inject(abort-current-inline-script.js, Promise)` for that domain.

**Comment by mapx-** [comment]
`meta-calculator.com`  ...is working fine.

**Comment by okiehsch** [comment]
Well, now I am officially without a clue what the problem is on your end 😕 .

**Comment by mapx-** [comment]
well, no big deal from my part, I arrived here from the other issue (receive-a-sms.com) and I wanted to test it ..

**Comment by okiehsch** [comment]
And I realised that I have to add another filter for Firefox users, so it was a productive confusion.

**Comment by gorhill** [comment]
@mapx- Can you put somewhere the raw content of `view-source:https://www.meta-chart.com/`? It's as if you are not served the same thing as us.

**Comment by okiehsch** [comment]
reference of the last [commit](https://github.com/uBlockOrigin/uAssets/commit/065a3967486dbb92901836253b0aa9524c6f4777)

**Comment by gotitbro** [comment]
@okiehsch @gorhill So filters not needed for Chromium users will be served to them as well? I believe there should be a different list for such issues specific to the workings of Firefox and enabled for default for the users of Firefox.

Just as was the need for uBO Extra for Chromium users.

**Comment by mapx-** [comment]
@gorhill here the raw content:
https://mega.nz/#!mUgWwAyY!9wtTaRbWaOKtX1C4d2h1pcA5AQRLAWxx5XuxnqMKVXY

**Comment by mapx-** [comment]
@gorhill  if you prefer see the site uploaded here, I still get the warning page
http://axistrivia.altervista.org/test/meta0.html

**Comment by okiehsch** [comment]
@mapx-

**Comment by mapx-** [comment]
No, I don't use Peter Lowe's list.
However, with / without that list the same result.

**Comment by okiehsch** [comment]
Well, without that list I can reproduce your issue and it also explains why `meta-calculator.com` does work on your end, because there is no connection to `limpingline.com` via `functionalclam.com` at that site.
I

**Comment by mapx-** [comment]
right. Now meta-chart is working fine, but ... my test page (axistrivia) still does not (even if `functionalclam` and `limpingline` are blocked)

**Comment by okiehsch** [comment]
No, your testpage also works if I block `limpingline` on my end.
If you go to `meta-chart.com` with `functionalclam` blocked, there should not be any `limpingline-request`
anyway.

**Comment by mapx-** [comment]
yes, (again) my fault (I'he commented the `promise` filter) for axistrivia.

On meta I can see only `functionalclam`, on axistrivia I see both:
![func](https://user-images.githubusercontent.com/10303732/31053176-a419bf42-a697-11e7-9e5c-5e5ae5948cdf.PNG)

**Comment by okiehsch** [comment]
Now we finally have identical results :)

**Comment by mapx-** [comment]
well, now I will enable peter lowe's list in all my browsers.

@okiehsch , 1 more thing: could you explain how you find the necessary numbers (300, 15000) in these filters ?
```
meta-chart.com##script:inject(setTimeout-defuser.js, "ubo", 300)
meta-chart.com##script:inject(setTimeout-defuser.js, [native code], 15000)
```
thank you

**Comment by okiehsch** [comment]
Well, `meta-chart.com` is a rather difficult example because of the sheer amount of setTimeout-calls.

1. Add `meta-chart.com##script:inject(setTimeout-logger.js)`
2. Disable `script:inject(abort-current-inline-script.js, Promise)`, because that filter disables some
`setTimeout-calls`.
3. Open your console and go to `meta-chart.com`

Now how you find the right parameter can be rather [tricky](https://github.com/uBlockOrigin/uAssets/commit/ce3e1e741aa4fd09097fd5c4ba043b715ffbd43b#comments).

And just fyi `example.com##script:inject(setTimeout-defuser.js)` defuses all setTmeout calls of that site.
